internal/datastore/proxy/caching_test.go

```diff
@@ -144,3 +148,47 @@ func TestSingleFlight(t *testing.T) {
 	dsMock.AssertExpectations(t)
 	oneReader.AssertExpectations(t)
 }
+
+type reader struct {
+	proxy_test.MockReader
+}
+
+func (r *reader) ReadNamespace(ctx context.Context, namespace string) (ns *core.NamespaceDefinition, lastWritten datastore.Revision, err error) {
+	time.Sleep(10 * time.Millisecond)
+	if errors.Is(ctx.Err(), context.Canceled) {
+		return nil, old, fmt.Errorf("error")
+	}
+	return &core.NamespaceDefinition{Name: namespace}, old, nil
+}
+
+func TestSingleFlightCancelled(t *testing.T) {
+	dsMock := &proxy_test.MockDatastore{}
+	ctx1, cancel1 := context.WithCancel(context.Background())
+	ctx2, cancel2 := context.WithCancel(context.Background())
+	defer cancel2()
+	defer cancel1()
+
+	dsMock.On("SnapshotReader", one).Return(&reader{MockReader: proxy_test.MockReader{}})
+
+	ds := NewCachingDatastoreProxy(dsMock, nil)
+
+	g := sync.WaitGroup{}
+	var ns2 *core.NamespaceDefinition
+	g.Add(2)
+	go func() {
+		_, _, _ = ds.SnapshotReader(one).ReadNamespace(ctx1, nsA)
+		g.Done()
+	}()
+	go func() {
+		time.Sleep(5 * time.Millisecond)
+		ns2, _, _ = ds.SnapshotReader(one).ReadNamespace(ctx2, nsA)
+		g.Done()
+	}()
+	cancel1()
+
+	g.Wait()
+	require.NotNil(t, ns2)
```

Comment:
this always fails in the previous implementation, because cancelling the first context poisons the singleflight response for the second request